src/libraries/System.Net.Security/src/System/Net/Security/TlsFrameHelper.cs

```diff
@@ -69,31 +72,85 @@ internal enum TlsAlertDescription : byte
+        ClientCertificateUrl = 2,
+        TrustedCaKeys = 3,
+        TruncatedHmac = 4,
+        CertificateStatusRequest = 5,
+        ApplicationProtocols = 16,
+        SupportedVersions = 43
+    }
+
     internal struct TlsFrameHeader
     {
         public TlsContentType Type;
         public SslProtocols Version;
         public int Length;
-    }
 
-    internal struct TlsFrameHandshakeInfo
-    {
-        public TlsFrameHeader Header;
-        public TlsHandshakeType HandshakeType;
-        public SslProtocols SupportedVersions;
-        public string? TargetName;
+        public override string ToString()
+        {
+            return $"{Version}:{Type}[{Length}]";
+        }
     }
 
     internal class TlsFrameHelper
     {
         public const int HeaderSize = 5;
 
+        [Flags]
+        public enum ProcessingOptions
+        {
+            All = 0,
+            ServerName = 1,
+            ApplicationProtocol = 2,
+            Versions = 4,
+        }
+
+        [Flags]
+        public enum ApplicationProtocolInfo
+        {
+            None = 0,
+            Http11 = 1,
+            Http2 = 2,
+            Http3 = 4,
+            Other = 128
+        }
+
+        public struct TlsFrameHandshakeInfo
+        {
+            public TlsFrameHeader Header;
+            public TlsHandshakeType HandshakeType;
+            public SslProtocols SupportedVersions;
+            public string TargetName;
+            public ApplicationProtocolInfo ApplicationProtocols;
+
+            public override string ToString()
+            {
+                return $"{Header.Version}:{HandshakeType}[{Header.Length}] TargetName='{TargetName}' SupportedVersion='{SupportedVersions}' ApplicationProtocols='{ApplicationProtocols}'";
+            }
+        }
+
+        public delegate bool HelloExtensionCallback(ref TlsFrameHandshakeInfo info, ExtensionType type, ReadOnlySpan<byte> extensionsData);
+
         private static byte[] s_protocolMismatch13 = new byte[] { (byte)TlsContentType.Alert, 3, 4, 0, 2, 2, 70 };
         private static byte[] s_protocolMismatch12 = new byte[] { (byte)TlsContentType.Alert, 3, 3, 0, 2, 2, 70 };
         private static byte[] s_protocolMismatch11 = new byte[] { (byte)TlsContentType.Alert, 3, 2, 0, 2, 2, 70 };
         private static byte[] s_protocolMismatch10 = new byte[] { (byte)TlsContentType.Alert, 3, 1, 0, 2, 2, 70 };
         private static byte[] s_protocolMismatch30 = new byte[] { (byte)TlsContentType.Alert, 3, 0, 0, 2, 2, 40 };
 
+        private const int UInt24Size = 3;
+        private const int RandomSize = 32;
+        private const int ProtocolVersionMajorOffset = 0;
+        private const int ProtocolVersionMinorOffset = 1;
+        private const int ProtocolVersionSize = 2;
+        private const int ProtocolVersionTlsMajorValue = 3;
+
+        private static readonly IdnMapping s_idnMapping = CreateIdnMapping();
+        private static readonly Encoding s_encoding = CreateEncoding();
```

Comment:
Do we need the helper functions?

Reply:
this came as pure copy from old SNI code. I can revisit that if you think we should.

Reply:
updated. Since this is single liner, it may be easier to read than jumping to helper.